How many late days are you using for this assignment? 

In [114]:
1

1

In [115]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.preprocessing import MinMaxScaler




# Part 1: Load the dataset

In [116]:
# Load the given datasets
dataC = pd.read_csv("chronic_kidney_disease_categorical.csv")
dataN = pd.read_csv("chronic_kidney_disease_numerical.csv")
# Print the data
pd.set_option("display.max.rows", 5)
dataC
dataN


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5
...,...,...,...,...,...,...,...,...,...,...,...,...
418,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9
419,926033,50.0,70.0,230.0,50.0,2.2,NaN,NaN,12.0,41.0,10400.0,4.6


# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: Each row has a unique ID attached to it so you can identify the individual cases. Then they each have a set of values under each column attached to that ID. In the numerical data CSV there is a lot of information that is given in the form of a number such as age, blood glucose random, and blood pressure (denoted age, bgr and bp in the column label). In the categorical dataset the IDs are put into categories, some binary some with more tags, and examples of this would be anemia (yes or no), hypertension (yes or no), and albumin (0,1,2, or 3). If an value is not present, meaning the dataset is incomplete, the place where the value would be is empty. 

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [117]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
print(dataN.duplicated(subset=['unique_id'], keep='first').sum())


# Drop these duplicate rows
dataN.drop_duplicates(subset=['unique_id'], keep='first')


# Repeat the same for categorical dataset. Print the duplicate rows and drop them
print(dataC.duplicated(subset=['unique_id'], keep='first').sum())
print(dataC.duplicated(subset=['unique_id'], keep='first'))
dataC.drop_duplicates(subset=['unique_id'], keep='first')


30
30
0      False
1      False
       ...  
428     True
429     True
Length: 430, dtype: bool


,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
399,273038,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
400,583971,1.0,0.0,NaN,normal,notpresent,notpresent,yes,no,no,good,yes,yes,ckd


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: Not all the unique ids are present in both datasets. In some rows, only the data from one of the data sets is filled out. The missing data is filled with NaN. 

In [118]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
pd.set_option("display.max.rows", 5)
merged = dataN.merge(dataC, how='outer')
#Print the combined dataset
merged


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
1,103570,24.0,80.0,125.0,NaN,NaN,136.0,3.5,15.4,43.0,5600.0,4.5,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
490,997629,35.0,NaN,99.0,30.0,0.5,135.0,4.9,15.4,48.0,5000.0,5.2,0.0,0.0,normal,normal,NaN,NaN,no,no,no,good,no,no,notckd
491,997876,64.0,70.0,107.0,15.0,NaN,NaN,NaN,12.8,38.0,NaN,NaN,0.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,no,ckd


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [119]:
# Calculate the percentage of rows that contain atleast one missing value
missing = merged.isnull().any(axis=1).sum()
# Print %
print(missing/len(merged))

# Drop these rows from the dataset
pd.set_option("display.max.rows", 5)
pruned = merged.dropna()
# Print the Dataset
pruned


0.5975609756097561


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
6,109053,64.0,60.0,106.0,27.0,0.7,150.0,3.3,14.4,42.0,8100.0,4.7,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
487,995177,69.0,70.0,214.0,96.0,6.3,120.0,3.9,9.4,28.0,11500.0,3.3,4.0,3.0,normal,abnormal,present,present,yes,yes,yes,good,yes,yes,ckd
488,996076,57.0,60.0,105.0,49.0,1.2,150.0,4.7,15.7,44.0,10400.0,6.2,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.4: Sort the dataset according to the Labels

In [120]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted = pruned.sort_values(by='Target')
# Print the dataset
pd.set_option("display.max.rows", 5)
sorted.index = list(range(1, len(sorted) + 1))
sorted

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
1,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
2,234299,48.0,80.0,95.0,163.0,7.7,136.0,3.8,9.8,32.0,6900.0,3.4,4.0,0.0,normal,abnormal,notpresent,notpresent,yes,no,no,good,no,yes,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
197,441170,79.0,80.0,111.0,44.0,1.2,146.0,3.6,16.3,40.0,8000.0,6.4,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
198,996076,57.0,60.0,105.0,49.0,1.2,150.0,4.7,15.7,44.0,10400.0,6.2,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [121]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
pd.set_option("display.max.rows", 5)
pd.set_option("display.max.columns", 36)
dummies = pd.get_dummies(sorted, columns=categorical_columns, dtype=int)
# Print the dataset
dummies

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
1,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1,3.0,0.0,0,1,1,0,1,0,1,0,0,1,1,0,1,0,0,1,0,1,0,1,1,0
2,234299,48.0,80.0,95.0,163.0,7.7,136.0,3.8,9.8,32.0,6900.0,3.4,4.0,0.0,0,1,1,0,1,0,1,0,0,1,1,0,1,0,1,0,1,0,0,1,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
197,441170,79.0,80.0,111.0,44.0,1.2,146.0,3.6,16.3,40.0,8000.0,6.4,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
198,996076,57.0,60.0,105.0,49.0,1.2,150.0,4.7,15.7,44.0,10400.0,6.2,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: This would be essentially equivelant and actually probably easier to read. This would make it much better if you had more than 2 values because having n amount of columns is needless clutter when things can easily be identified by their n tag. 

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [122]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
for column in numerical_columns:
    mean = dummies[column].mean()
    std_dev = dummies[column].std()
    lower_bound = mean - 3 * std_dev
    upper_bound = mean + 3 * std_dev
    no_outliers = dummies[(dummies[column] >= lower_bound) & (dummies[column] <= upper_bound)]

# Print the dataset
no_outliers

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
1,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1,3.0,0.0,0,1,1,0,1,0,1,0,0,1,1,0,1,0,0,1,0,1,0,1,1,0
2,234299,48.0,80.0,95.0,163.0,7.7,136.0,3.8,9.8,32.0,6900.0,3.4,4.0,0.0,0,1,1,0,1,0,1,0,0,1,1,0,1,0,1,0,1,0,0,1,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
197,441170,79.0,80.0,111.0,44.0,1.2,146.0,3.6,16.3,40.0,8000.0,6.4,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
198,996076,57.0,60.0,105.0,49.0,1.2,150.0,4.7,15.7,44.0,10400.0,6.2,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [123]:
# Normalize the all Numerical Attributes in the dataset.
scaler = MinMaxScaler()

no_outliers.loc[:, numerical_columns] = scaler.fit_transform(no_outliers[numerical_columns])

# Print the dataset
no_outliers

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
1,102677,0.688312,0.333333,0.014286,0.588629,0.986486,0.615385,0.114607,0.272109,0.288889,0.000000,0.000000,3.0,0.0,0,1,1,0,1,0,1,0,0,1,1,0,1,0,0,1,0,1,0,1,1,0
2,234299,0.545455,0.500000,0.059524,0.511706,0.493243,0.641026,0.029213,0.455782,0.511111,0.137168,0.295455,4.0,0.0,0,1,1,0,1,0,1,0,0,1,1,0,1,0,1,0,1,0,0,1,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
197,441170,0.948052,0.500000,0.097619,0.113712,0.054054,0.897436,0.024719,0.897959,0.688889,0.185841,0.977273,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
198,996076,0.662338,0.166667,0.083333,0.130435,0.054054,1.000000,0.049438,0.857143,0.777778,0.292035,0.931818,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: The only one I can think of removing would be the unique ID. at this point we have combined the data and the only reason we would need it is for indexing. We already have an index, it doesn't matter who these people are we are just using anonymous bulk data at this point for modeling. 

In [124]:
#Remove that column
dropped = no_outliers.drop(columns='unique_id')

# Print the dataset
dropped

,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
1,0.688312,0.333333,0.014286,0.588629,0.986486,0.615385,0.114607,0.272109,0.288889,0.000000,0.000000,3.0,0.0,0,1,1,0,1,0,1,0,0,1,1,0,1,0,0,1,0,1,0,1,1,0
2,0.545455,0.500000,0.059524,0.511706,0.493243,0.641026,0.029213,0.455782,0.511111,0.137168,0.295455,4.0,0.0,0,1,1,0,1,0,1,0,0,1,1,0,1,0,1,0,1,0,0,1,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
197,0.948052,0.500000,0.097619,0.113712,0.054054,0.897436,0.024719,0.897959,0.688889,0.185841,0.977273,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
198,0.662338,0.166667,0.083333,0.130435,0.054054,1.000000,0.049438,0.857143,0.777778,0.292035,0.931818,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [125]:
# Export the dataframe to a new csv file
dropped.to_csv('CKD_data.csv', index=False)
# Export the dataframe to a new json file
dropped.to_json('data.json')
pd.set_option("display.max.rows", 15)

dropped.head(15)

,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
1,0.688312,0.333333,0.014286,0.588629,0.986486,0.615385,0.114607,0.272109,0.288889,0.000000,0.000000,3.0,0.0,0,1,1,0,1,0,1,0,0,1,1,0,1,0,0,1,0,1,0,1,1,0
2,0.545455,0.500000,0.059524,0.511706,0.493243,0.641026,0.029213,0.455782,0.511111,0.137168,0.295455,4.0,0.0,0,1,1,0,1,0,1,0,0,1,1,0,1,0,1,0,1,0,0,1,1,0
3,0.571429,0.666667,0.138095,0.662207,0.594595,0.589744,0.051685,0.346939,0.288889,0.553097,0.136364,2.0,0.0,0,1,1,0,0,1,0,1,1,0,1,0,1,0,0,1,0,1,0,1,1,0
4,0.610390,0.666667,0.000000,0.324415,0.459459,0.076923,0.026966,0.435374,0.444444,0.367257,0.363636,2.0,0.0,1,0,1,0,0,1,1,0,0,1,0,1,1,0,0,1,1,0,0,1,1,0
5,0.701299,0.166667,0.519048,0.086957,0.087838,0.487179,0.011236,0.326531,0.355556,0.504425,0.204545,3.0,1.0,0,1,1,0,0,1,1,0,0,1,1,0,1,0,0,1,1,0,0,1,1,0
6,0.870130,0.333333,0.000000,0.073579,0.033784,0.358974,0.033708,0.469388,0.444444,0.668142,0.318182,0.0,0.0,0,1,0,1,1,0,1,0,0,1,0,1,1,0,1,0,0,1,1,0,1,0
7,0.194805,0.666667,0.088095,0.100334,0.087838,0.358974,0.022472,0.353741,0.311111,0.380531,0.409091,4.0,0.0,0,1,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,0,1,1,0
8,0.844156,0.166667,0.114286,0.384615,0.331081,0.641026,0.053933,0.564626,0.577778,0.504425,0.500000,4.0,0.0,0,1,0,1,1,0,1,0,0,1,0,1,1,0,0,1,0,1,1,0,1,0
9,0.844156,0.333333,0.354762,0.240803,0.216216,0.564103,0.042697,0.496599,0.533333,0.079646,0.340909,3.0,0.0,0,1,1,0,0,1,0,1,0,1,0,1,0,1,1,0,1,0,1,0,1,0
10,0.805195,0.500000,0.207143,0.267559,0.250000,0.487179,0.087640,0.170068,0.155556,0.318584,0.113636,3.0,2.0,0,1,1,0,0,1,0,1,0,1,0,1,0,1,0,1,0,1,1,0,1,0


# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

Paste here: i could only get it to do some at a time this is terrible |    |      age |        bp |       bgr |        bu |        sc |        sod |        pot |      hemo |       pcv |      wbcc |      rbcc |    al |   su |   rbc_abnormal |   rbc_normal |   pc_abnormal |   pc_normal |   pcc_notpresent |   pcc_present |   ba_notpresent |   ba_present |   htn_no |   htn_yes |   dm_no |   dm_yes |   cad_no |   cad_yes |   appet_good |   appet_poor |   pe_no |   pe_yes |   ane_no |   ane_yes |   Target_ckd |   Target_notckd |
|----|----------|-----------|-----------|-----------|-----------|------------|------------|-----------|-----------|-----------|-----------|-------|------|----------------|--------------|----------------|-------------|------------------|----------------|----------------|--------------|----------|-----------|---------|----------|----------|-----------|--------------|---------------|---------|----------|----------|-----------|---------------|------------------|
|  6 | 0.870130 | 0.333333  | 0.000000  | 0.073579  | 0.033784  | 0.358974   | 0.033708   | 0.469388  | 0.444444  | 0.668142  | 0.318182  | 0.0   | 0.0  |              0 |            1 |              0 |           1 |                1 |              0 |              1 |            0 |        0 |         1 |       0 |        1 |        1 |         0 |            1 |             0 |       1 |        0 |        1 |         1 |              0 |                0 |
|  7 | 0.194805 | 0.666667  | 0.088095  | 0.100334  | 0.087838  | 0.358974   | 0.022472   | 0.353741  | 0.311111  | 0.380531  | 0.409091  | 4.0   | 0.0  |              0 |            1 |              1 |           0 |                0 |              1 |              0 |            1 |        1 |         0 |       1 |        0 |        1 |         0 |            1 |             0 |       1 |        0 |        1 |         1 |              0 |                0 |
|  8 | 0.844156 | 0.166667  | 0.114286  | 0.384615  | 0.331081  | 0.641026   | 0.053933   | 0.564626  | 0.577778  | 0.504425  | 0.500000  | 4.0   | 0.0  |              0 |            1 |              0 |           1 |                1 |              0 |              1 |            0 |        0 |         1 |       0 |        1 |        1 |         0 |            1 |             0 |       1 |        0 |        1 |         1 |              0 |                0 |
|  9 | 0.844156 | 0.333333  | 0.354762  | 0.240803  | 0.216216  | 0.564103   | 0.042697   | 0.496599  | 0.533333  | 0.079646  | 0.340909  | 3.0   | 0.0  |              0 |            1 |              1 |           0 |                0 |              1 |              0 |            1 |        0 |         1 |       0 |        1 |        1 |         0 |            1 |             0 |       1 |        0 |        1 |         1 |              0 |                0 |
| 10 | 0.805195 | 0.500000  | 0.207143  | 0.267559  | 0.250000  | 0.487179   | 0.087640   | 0.170068  | 0.155556  | 0.318584  | 0.113636  | 3.0   | 2.0  |              0 |            1 |              1 |           0 |                0 |              1 |              0 |            1 |        0 |         1 |       0 |        1 |        1 |         0 |            1 |             0 |       1 |        0 |        1 |         1 |              0 |                0 |


### Paste the markdown table here

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [126]:
# Code to rename all the columns in the dataset
column_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}

# Rename columns
dropped.rename(columns=column_mapping, inplace=True)

dropped


,age,blood pressure,blood glucose random,blood urea,serum creatinine,sodium,potassium,hemoglobin,packed cell volume,wbcc,rbcc,albumin,sugar,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
1,0.688312,0.333333,0.014286,0.588629,0.986486,0.615385,0.114607,0.272109,0.288889,0.000000,0.000000,3.0,0.0,0,1,1,0,1,0,1,0,0,1,1,0,1,0,0,1,0,1,0,1,1,0
2,0.545455,0.500000,0.059524,0.511706,0.493243,0.641026,0.029213,0.455782,0.511111,0.137168,0.295455,4.0,0.0,0,1,1,0,1,0,1,0,0,1,1,0,1,0,1,0,1,0,0,1,1,0
3,0.571429,0.666667,0.138095,0.662207,0.594595,0.589744,0.051685,0.346939,0.288889,0.553097,0.136364,2.0,0.0,0,1,1,0,0,1,0,1,1,0,1,0,1,0,0,1,0,1,0,1,1,0
4,0.610390,0.666667,0.000000,0.324415,0.459459,0.076923,0.026966,0.435374,0.444444,0.367257,0.363636,2.0,0.0,1,0,1,0,0,1,1,0,0,1,0,1,1,0,0,1,1,0,0,1,1,0
5,0.701299,0.166667,0.519048,0.086957,0.087838,0.487179,0.011236,0.326531,0.355556,0.504425,0.204545,3.0,1.0,0,1,1,0,0,1,1,0,0,1,1,0,1,0,0,1,1,0,0,1,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
194,0.571429,0.500000,0.064286,0.100334,0.013514,1.000000,0.044944,0.755102,0.866667,0.296460,0.659091,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
195,0.077922,0.500000,0.071429,0.053512,0.013514,0.666667,0.042697,0.863946,0.888889,0.123894,0.750000,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
196,0.948052,0.500000,0.097619,0.113712,0.054054,0.897436,0.024719,0.897959,0.688889,0.185841,0.977273,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
197,0.948052,0.500000,0.097619,0.113712,0.054054,0.897436,0.024719,0.897959,0.688889,0.185841,0.977273,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [129]:
# Converted SQL to Pandas code
# Example DataFrame
data = {
    'Target': ['ckd', 'notckd', 'ckd', 'ckd', 'notckd', 'notckd', 'ckd'],
}

df = pd.DataFrame(data)

# Equivalent pandas query
grouped_result = df.groupby('Target').size().reset_index(name='count')

# Print the result
print(grouped_result)

   Target  count
0     ckd      4
1  notckd      3


*Describe what the above code does here*

it made an example table and converted it and printed it as it would for a panda file.